File: src/src/WrapperCCService.js
const fs = require("fs");
const bindings = require('bindings');
const HttpQueryService = require("./HttpQueryService");

/** Implements a wrapper to C/C++ implemented functions.
 *
 * - A POST service is attached to the given route.
 * 
 * - This is the server side of a simple [client](./WrapperCC.html)/server  C/C++ wrapping mechanism.
 *
 * - The local service is attached to an express web mechanism, using a construct of the form:
 *```
 * // Loads required packages
 * const express = require("express");
 * const server = express();
 * const CCWrapperService = require("./CCWrapper.js");
 * 
 * // Defines a minimal persistantdata service
 * new WrapperCCService(server, "/_ccwrapper");
 * server.listen(8080);
 *```
 *
 * - The C/C++ wrapping mechanism is implemented vis a unique C [`wrap(query)`](global.html) function, stored in a `wrap.cpp` file in the `./src` directory.
 *
 * @param {express} server      The express server mechanism.
 * @param {string}  route       The route to this service.
 * @param {string}  target_name The wrapper name as defined in `binding.gyp` 
 * ```
 * {
 *   "targets": [{
 *     "target_name": "wrapper",
 * ../..
 * ```
 * @class
 */
var WrapperCCService = function(server, route = "/_ccwrapper", target_name = "wrapper") {
  // Dumps all available info on error
  const errorLog = function(what) {
    console.log("Error in CCWrapperService: the wrapper is " + what);
    onError = what;
  }
  var onError = "";
  // Implements the wrapper if available
  if (fs.existsSync("./build/Release/" + target_name + ".node")) {
    var wrapper = bindings({
      bindings: target_name,
      module_root: process.cwd()
    });
    if (wrapper == undefined)
      errorLog("not loadable");
  } else
    errorLog("not available");
  // Implements the service
  const doQuery = function(query, answer) {
    // Prepares the query to be compatible with C/C++ wrapper string value
    for (let name in query) {
      if (typeof query[name] != "string") {
        query[name] = "" + query[name];
      }
    }
    if (onError == "") {
      answer(wrapper.wrap(query));
    } else {
      answer("{ \"onError\": \"wrapper " + onError + "\", \"query\": \"" + JSON.stringify(query) + "\" }");
    }
  };
  // Connects to a Http query service
  new HttpQueryService(server, route, doQuery);
};

module.exports = WrapperCCService;

File: src/src/WrapperCCService.cpp
#include <napi.h>

#include <WrapperCCService.hpp>

std::string nowrap(const std::map < std::string, std::string > &query)
{
  std::string string = "";
  for(auto item : query) {
    string += (string == "" ? "" : "&") + item.first + "=" + item.second;
  }
  return "{ \"error\": \"wrong or unimplemented query\", \"query\": \"" + string + "\" }";
}
// Implements the wrapping mechanism of the wrap() function
Napi::String NapiWrapperMethod(const Napi::CallbackInfo& info)
{
  std::map < std::string, std::string > query;
  {
    Napi::Object values = info[0].As < Napi::Object > ();
    Napi::Array names = values.GetPropertyNames();
    for(unsigned int i = 0; i < names.Length(); i++) {
      std::string name = names.Get(i).As < Napi::String > ().Utf8Value();
      query[name] = values.Get(names.Get(i)).As < Napi::String > ().Utf8Value();
    }
  }
  Napi::String res = Napi::String::New(info.Env(), wrap(query));
  return res;
}
// Registers the wrapping mechanism of the C wrap() function as a `wrap` JS function
Napi::Object NapiWrapperInit(Napi::Env env, Napi::Object exports)
{
  exports.Set(Napi::String::New(env, "wrap"), Napi::Function::New(env, NapiWrapperMethod));
  return exports;
}
// Registers the binding as a `wrapper` module
NODE_API_MODULE(wrapper, NapiWrapperInit)
